### ISSUE TITLE
Ability to zip columns together when joining

### ISSUE THREAD

**Issue by SCVFlare** [OpenRefine/OpenRefine#5440]
It would be easier if OpenRefine had a zip-like function when joining columns with multi-valued cells. 
Imagine the following table :
| Col1 | Col2 | Col3 |
|--|--|--|
| 1,2,3 | A,B,C | 1A,2B,3C |

Col3 is the desired outcome as then I can split the cell on `,` to achieve my final wish:

| Col1 |
|--|
| 1A |
| 2B |
| 3C |

The idea is to have this in the UI in the join columns menu as I have encountered this issue in several datasets

### Proposed solution

To achieve this I made a GREL expression:
```
forEachIndex(
	cells["Col 1"].value.split(","),
	i,
	v,
	v + cells["Col 2"].value.split(",").get(i)
).join(",")
```
However, this is error-prone and I want the python-like behavior where the zipping stops when one of the arrays runs out of elements. If I do a check things become much more ugly and increasing the amount of columns I want to zip becomes very complex.

```
if(
	length(cells["Col 1"].value.split(","))>length(cells["Col 2"].value.split(",")),
	forEachIndex(
		cells["Col 1"].value.split(",").slice(0,length(cells["Col 2"].value.split(","))),
		i,
		v,
		v + cells["Col 2"].value.split(",").get(i)
	).join(","),
	forEachIndex(
		cells["Col 1"].value.split(","),
		i,
		v,
		v + cells["Col 2"].value.split(",").get(i)
	).join(",")
)
```
### Alternatives considered

I assume that using jython for the purpose would be much easier as python has a zip function but it would be nice to have it in the UI.

**Comment by wetneb** [comment]
I think it is legitimate to ask for a `zip` function in GREL indeed (but as you point out it's much easier to use a fully-fledged programming language with Jython).

I don't think we can expose every single GREL function in the UI - especially here because it is something that works on arrays (which cannot be stored in cells directly), so you would need to prompt the user for the appropriate separators and column names… Let's keep this issue for implementing a `zip` GREL function like in Python though.

**Comment by jquartel** [comment]
Interestingly, this would have worked with my proposal (#4185) to overload some operators to work element-wise with arrays. Although my own intentions were to be able to work with numerical arrays (e.g. [1,2] + [0,1] to give [1,3]), I showed an example for for strings too (so you would just use + on your two split arrays to 'zip' them). However, it was deemed more logical to expect + to do something like concatenate the arrays (for which we currently have no function, as far as I can tell).

**Comment by Sudo-Aryan** [comment]
Can I be assigned to this? 

Regards Aryan

**Comment by DiamondMyK** [comment]
Hi, I am a junior undergraduate student and I want to make some contributions to the open source community. I'm happy to deal with this issue. Can you assign it to me? Thank you! @wetneb

**Comment by DiamondMyK** [comment]
Could you please give me some preliminary suggestions? I would really appreciate it!

**Comment by DiamondMyK** [comment]
Thank you very much for your reply and help! I will try my best to solve this problem!

**Comment by AbhishekSharma55** [comment]
Hi there,

I want to contribute to this project and take on this issue as my first task. Could you please assign it to me? I am excited to work towards resolving it and will keep the team updated on my progress.

Thanks

**Comment by sunilnatraj** [comment]
@wetneb I want to validate my understanding of this feature request. 

The zip function will be used in the "add column based on this column" user can specify the columns to zip and separator

GREL to zip columns a1, b2, c3 with separator as comma
[ row.record.cells["a1"].value[0], row.record.cells["b2"].value[0], row.record.cells["c3"].value[0] ].zip(",")

**Comment by tfmorris** [comment]
Python's `zip` doesn't actually do what's described above. Given `x=[1,2,3]` and y=`['a','b','c']` then `zip(x,y)` produces ((1,'a'),(2,'b'),(3,'c')), not ('1a','2b','3c'). It's up to the consumer of the new iterable to decide what to do with the tuples generated.

I think the spec should be clarified for this to be considered a Good First Issue.

**Comment by sunilnatraj** [comment]
@tfmorris The output of the zip function will be as below


Col1 | Col2 | Col3
-- | -- | --
1,2,3 | A,B,C | 1A,2B,3C
1,2,3 | A,B | 1A,2B
1,2 | A,B,C | 1A,2B

**Comment by ostephens** [comment]
The python implementation as highlighted by @tfmorris seems more flexible than the proposed solution and would allow for outcomes other than concatenating pieces of text in two arrays. It would, on the other hand, make it more work for the user in the case of them wanting the concatenated text.

**As a starting point for discussion only** .... a proposed spec:

- New GREL function `zip()`
- `zip()` should require two or more parameters, each of which should be an array or arrayNode (the latter to deal with parsed JSON)
- the outcome of `zip()` should be an array of arrays, each inner array containing the same positioned value from the input arrays
- if the input arrays to the `zip()` are not of equal length, `zip()` should work to the length of the longest array, filling any missing values with `null` (this is different to the Python zip with stops when it reaches the length of the shortest of the input arrays, but it feels more in keeping with the typical behaviour users have asked from OR in relation to similar situations like joining columns together)
```
zip([1,2],[3,4]) -> [[1,3],[2,4]]
zip([1,2],[3,4],[5,6]) -> [[1,3,5],[2,4,6]]
zip([1,2,3],[4,5]) ->  [[1,4],[2,5],[3,null]]

**Comment by sunilnatraj** [comment]
@ostephens A point that @wetneb mentioned above is that an Array cannot be saved in a cell.

**Comment by ostephens** [comment]
@sunilnatraj yes that's true but we have many GREL functions that result in an array e.g. `split()`, `find()`, `match()`, `filter()` etc.. The user then has to do some further step to process the array. This is not an unusual operation for OpenRefine although it does mean the user has to do the work

**Comment by sunilnatraj** [comment]
@ostephens The initial requirement was to have the column data zipped and saved in a column achieving this may be not easy with above zip function definition, should we support 2 variants of the zip function?

**Comment by jquartel** [comment]
I think we only need the one function, as described by @ostephens. The remaining task to join the array elements in a string concatenation can be achieved simply enough with a subsequent application of join(). For the stated example, I think building Col3 would look something like

forEach(forEach(["Col1","Col2"],c,cells[c].value.split(",")).zip(), cz, cz.join("")).join(",")

Leaving the function general allows it to be used for non-string operations (like mathematical ones which I'm interested in 😉 ).

**Comment by ostephens** [comment]
@sunilnatraj in response to the original request for a zip function on doing a column join @wetneb said:

>I don't think we can expose every single GREL function in the UI - especially here because it is something that works on arrays (which cannot be stored in cells directly), so you would need to prompt the user for the appropriate separators and column names… Let's keep this issue for implementing a zip GREL function like in Python though.

I was taking this as my starting point for my proposal above.

**Comment by ostephens** [comment]
@jquartel said:
> For the stated example, I think building Col3 would look something like `forEach(forEach(["Col1","Col2"],c,cells[c].value.split(",")).zip(), cz, cz.join("")).join(",")`

Agreed. This is definitely not simple :) tbf if I then compare to the original posters option:
```
forEachIndex(
	cells["Col 1"].value.split(","),
	i,
	v,
	v + cells["Col 2"].value.split(",").get(i)
).join(",")
```
it feels more complex rather than less complex!

This is slightly swaying me away from usefulness of zip tbh!

**Comment by ostephens** [comment]
The original poster says:
> I want the python-like behavior where the zipping stops when one of the arrays runs out of elements. If I do a check things become much more ugly and increasing the amount of columns I want to zip becomes very complex.

I said:
> if the input arrays to the zip() are not of equal length, zip() should work to the length of the longest array, filling any missing values with null

I'm inclined to reverse this part of my proposal based on this comment from the OP which I missed in my first read through

**Comment by jquartel** [comment]
> This is slightly swaying me away from usefulness of zip tbh!

I agree it's still not completely simple, but it does remove the need for the indexing and may improve error detection (which appears to be one of the original concerns).

Of course there is also my old suggestion to overload some operators to work element-wise on arrays. So if it were possible to do, say, ["a","b"]+["1","2"] to give ["a1","b2"] then the solution could look like

`[cells.Col1.value.split(",")+cells.Col2.value.split(",")].join(",")`

but this might still get complicated if there are many columns to zip together.

If we are primarily concerned with strings, then maybe a new multi-array function like joinEach will do:

`forEach(["Col1","Col2"],c,cells[c].value.split(",")).joinEach("").join(",")`

So joinEach([a1, a2, ...],sep) turns an array of arrays of strings into an array of (element-wise) joined strings.

**Comment by sunilnatraj** [comment]
@ostephens Can you let me know if the zip function is final.  Based on the updated definition PFB working sample

**Inputs to Zip function** 
- Array of strings [Bob has a cat & dog, Cat's name is Dotty, Dog's name is Dots]
- Array of integers [11, 22, 33, 44, 55]
- ArrayNode {"objects": ["Doe", "Foe", 456, 789]}

**Output**
[[Bob has a cat & dog, 11, "Doe"], [Cat's name is Dotty, 22, "Foe"], [Dog's name is Dots, 33, 456]]

**Comment by ostephens** [comment]
@sunilnatraj this seems OK to me and it seems that @jquartel agrees. It would be good to get a thumbs up (or not) from at least one of @tfmorris @wetneb. I don't know if the OP is still working with OpenRefine but will also tag @SCVFlare for comment

**Comment by tfmorris** [comment]
I don't understand the "PFB" reference, but the example looks good to me. Thank you @sunilnatraj @ostephens @jquartel for driving the discussion forward.

The request for "zip-like" doesn't mean that we need to slavishly follow Python's example, but I do think it provides a useful and familiar reference point for folks. In reference to @ostephens earlier request that the maximal length array control iteration, Python also has [`zip_longest`](https://docs.python.org/3/library/itertools.html#itertools.zip_longest) which we could implement, if desired (but I'm not trying to extend this issue to cover that case).

To cover a few additional cases, I'd also say that:
- nulls get passed through to the output, not elided as in some cases of OpenRefine's variable handling of this case
- array data types can include anything supported by OpenRefine including dates (ie datetimes), and any type of number (float & integer)
- mixed array types are allowed (as the `ArrayNode` example above demonstrates)

**Comment by sunilnatraj** [comment]
@wetneb Can you assign this to me

@wetneb @tfmorris @ostephens Raised PR - https://github.com/OpenRefine/OpenRefine/pull/6739

**Comment by sunilnatraj** [comment]
@ostephens @tfmorris Logged issue for Zip function documentation - https://github.com/OpenRefine/openrefine.org/issues/369
